contracts/strategies/DonationVotingOnchain.sol

```diff
@@ -0,0 +1,269 @@
+
+        emit Initialized(_poolId, _data);
+    }
+
+    /// ===============================
+    /// ======= External/Custom =======
+    /// ===============================
+
+    /// @notice Sets the start and end dates.
+    /// @dev The 'msg.sender' must be a pool manager.
+    /// @param _registrationStartTime The start time for the registration
+    /// @param _registrationEndTime The end time for the registration
+    /// @param _allocationStartTime The start time for the allocation
+    /// @param _allocationEndTime The end time for the allocation
+    function updatePoolTimestamps(
+        uint64 _registrationStartTime,
+        uint64 _registrationEndTime,
+        uint64 _allocationStartTime,
+        uint64 _allocationEndTime
+    ) external onlyPoolManager(msg.sender) {
+        if (_allocationStartTime > _allocationEndTime) revert INVALID();
+        allocationStartTime = _allocationStartTime;
+        allocationEndTime = _allocationEndTime;
+        emit AllocationTimestampsUpdated(allocationStartTime, allocationEndTime, msg.sender);
+
+        _updatePoolTimestamps(_registrationStartTime, _registrationEndTime);
+    }
+
+    /// @notice Sets recipient statuses.
+    /// @param statuses new statuses
+    /// @param refRecipientsCounter the recipientCounter the transaction is based on
+    function reviewRecipients(ApplicationStatus[] memory statuses, uint256 refRecipientsCounter)
+        public
+        override
+        onlyActiveRegistration
+    {
+        super.reviewRecipients(statuses, refRecipientsCounter);
+    }
+
+    /// ====================================
+    /// ============ Internal ==============
+    /// ====================================
+
+    /// @notice This will allocate to recipients.
+    /// @param _recipients The addresses of the recipients to allocate to
+    /// @param _amounts The amounts to allocate to the recipients
+    /// @param _sender The address of the sender
+    function _allocate(address[] memory _recipients, uint256[] memory _amounts, bytes memory, address _sender)
+        internal
+        virtual
+        override
+        onlyActiveAllocation
+    {
+        uint256 totalAmount;
+        for (uint256 i = 0; i < _recipients.length; i++) {
+            if (!_isAcceptedRecipient(_recipients[i])) revert RECIPIENT_ERROR(_recipients[i]);
+
+            // Update the total payout amount for the claim and the total claimable amount
+            amountAllocated[_recipients[i]] += _amounts[i];
+            totalAmount += _amounts[i];
+
+            emit Allocated(_recipients[i], _sender, _amounts[i], abi.encode(allocationToken));
+        }
+
+        if (allocationToken == NATIVE) {
+            if (msg.value != totalAmount) revert AMOUNT_MISMATCH();
+        } else {
+            SafeTransferLib.safeTransferFrom(allocationToken, _sender, address(this), totalAmount);
+        }
+
+        QFState.fund(_recipients, _amounts);
```

Comment:
We should emit an event after the allocation

Reply:
It's emitted per recipient https://github.com/allo-protocol/allo-v2/blob/6aa3261364b86c4540d199419af9b3550fa72e8b/contracts/strategies/DonationVotingOnchain.sol#L184-L192